BZ-17985 Added Toast Component

badri-singhal opened the pull request:
Created a toast component, with different types like success, error, info and warn.

Dev Proof:- https://drive.google.com/file/d/1xEWSKZvXpi7DW_lPqi5bOvpOMb1snrqB/view?usp=sharing

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,149 @@
+  
+    let visible = false;
+  
+    $: leftIcon = properties.leftIcon ? properties.leftIcon : '';
```
not needed to be reactive

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/properties.ts:
```diff
@@ -0,0 +1,23 @@
+export const defaultToastProperties: ToastProperties = {
+    type: 'info',
+    message: 'Toast',
+    backgroundColor: 'green',
```
why type info and keeping background color green in defaultToastProperties?

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,149 @@
+      height: auto;
+      width: auto;
+      align-items: center;
+      margin: 0px 10px 10px 10px;
```
Add css variables

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,149 @@
+      {#if properties.showRightIcon}
+        <div>
+          <!-- svelte-ignore a11y-click-events-have-key-events -->
+          <!-- svelte-ignore a11y-no-static-element-interactions -->
```
with proper on:keydown one warning can be removed

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,149 @@
+  
+    onMount(() => {
+      visible = true
+      setTimeout(() => {
```
collect the timeOut response and clear it on onDestroy

badri-singhal replied:
updated.

swaroopvarma1 reviewed (commented):
Resolve comments

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,157 @@
+<script lang="ts">
+    import { fly } from 'svelte/transition';
+  
```
fix spacing

swaroopvarma1 replied:
run prettier

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,157 @@
+    export let properties: ToastProperties = defaultToastProperties;
+  
+    let visible = false;
+    let leftIcon = properties.leftIcon ? properties.leftIcon : '';
```
let leftIcon = properties.leftIcon ?? null

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,157 @@
+    function getTypeClass(type: string): string {
+      switch (type) {
+        case 'success':
+          if (leftIcon.length == 0) {
```
add check on null

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,157 @@
+          }
+          return 'warn-toast';
+        default:
+          return '';
```
return null

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,157 @@
+    .toast {
+      padding: 10px;
+      font-size: 14px;
+      height: var(--toast-height);
```
height: var(--toast-height, auto);

badri-singhal replied:
updated.

swaroopvarma1 commented:
Add dev proof

badri-singhal commented:
> Add dev proof

Added

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,154 @@
+    style="border-radius: {properties.borderRadius};"
+    in:fly={{ x: 500, y: 0, duration: 400 }}
+    out:fly={{ x: 500, y: 0, duration: 800 }}
+    on:outroend={() => dispatch('kill')}
```
move this logic to a function and dispatch from function itself

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,154 @@
+      <div>
+        <div
+          class="close-button"
+          on:click={() => {
```
move this to a function as well

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,154 @@
+<script lang="ts">
+  import { fly } from 'svelte/transition';
+
```
remove extra line space here

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,154 @@
+    ...properties
+  };
+
+  async function hideToast() {
```
why async?

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,154 @@
+  export let properties: ToastProperties = defaultToastProperties;
+  const dispatch = createEventDispatcher();
+
+  let visible = false;
```
showToast

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,154 @@
+    clearInterval(timeout);
+  });
+
+  function getToastType(type: string): string | null {
```
return enum toast Type  here instead of string

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,159 @@
+
+  export let properties: ToastProperties = defaultToastProperties;
+
+  enum ToastType {
```
no need to create a new enum here, you can use already created type ToastType which was exported in properties.ts

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,159 @@
+  import { defaultToastProperties, type ToastProperties } from './properties';
+  import errorIcon from './error.svg';
+  import closeIcon from './close.svg';
+  import infoIcon from './info.svg';
```
send these icons via slack, will upload them

badri-singhal replied:
done.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,159 @@
+    style="border-radius: {properties.borderRadius};"
+    in:fly={{ x: 500, y: 0, duration: 400 }}
+    out:fly={{ x: 500, y: 0, duration: 800 }}
+    on:outroend={dispatchKill}
```
handleAnimationEnd

swaroopvarma1 replied:
dispatchKill -> handleAnimationEnd

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,153 @@
+<style>
+  .toast {
+    padding: 10px;
+    font-size: 14px;
```
add css variables

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,153 @@
+  .close-button {
+    width: 20px;
+    height: 20px;
+    cursor: pointer;
```
css variable for this too

badri-singhal replied:
updated.

sinha-sahil commented on src/lib/Toast/properties.ts:
```diff
@@ -0,0 +1,23 @@
+  message: string;
+  backgroundColor?: string;
+  textColor?: string;
+  leftIcon?: string | null;
```
Lets use `?` or `null`. One of them are sufficient to indicate optional params

badri-singhal replied:
updated.

sinha-sahil commented on src/lib/Toast/properties.ts:
```diff
@@ -0,0 +1,23 @@
+export const defaultToastProperties: ToastProperties = {
+  type: 'success',
+  message: 'Toast',
+  backgroundColor: '#24AA5A',
```
All styling parameters must be passed in css variables.

Lets add primitive classes available for `success`, `failure` & `warning` in css.
The same class can be applied based upon type of the toast.

badri-singhal replied:
updated.

sinha-sahil commented on src/lib/Toast/properties.ts:
```diff
@@ -0,0 +1,23 @@
+export type ToastType = 'success' | 'error' | 'info' | 'warn';
+
+export type ToastProperties = {
+  type: ToastType;
```
lets make type an optional param. users can avoid passing type in case they want to show some kind of custom message.

badri-singhal replied:
updated.

sinha-sahil commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  .success {
+    background-color: #24aa5a;
```
all styling must be exposed via css variables

badri-singhal replied:
updated.

sinha-sahil commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,153 @@
+  let leftIcon = properties.leftIcon ?? null;
+  let timeout: ReturnType<typeof setInterval>;
+
+  properties = {
```
No need of doing this. 
If properties is exported, then passed params will be received in the same.

badri-singhal replied:
updated.

sinha-sahil commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  function handleAnimationEnd() {
+    dispatch('kill');
```
Sounds weird. Name the events dispatched properly.

badri-singhal replied:
updated.

sinha-sahil commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,153 @@
+    switch (type) {
+      case 'success':
+        if (!leftIcon) {
+          leftIcon = 'https://sdk.breeze.in/gallery/icons/order-success-icon.svg';
```
We should avoid hard coding icons here. 
Users might not want to show icon if `leftIcon` was passed as falsey value.

badri-singhal replied:
updated.

sishir2001 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,122 @@
+{#if showToast}
+  <div
+    class="toast"
+    style="background-color: var(--toast-background-color, {backgroundColor})"
```
@badri-singhal  Can these styles be inside toast class?

badri-singhal replied:
we have to consider the below cases and for that we need a variable.

- If background color is passed then apply that background
- If it is not passed then apply on the basis of type
- If type is also not passed then apply a default info background color.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  function handleAnimationEnd() {
+    dispatch('handleAnimationEnd');
```
dispatch('onToastHide') or something related

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,122 @@
+    width: var(--toast-width, 20px);
+    height: var(--toast-height, 20px);
+    cursor: var(--toast-cursor, pointer);
+    gap: 6px;
```
add css variables to all style values here

badri-singhal replied:
updated.

swaroopvarma1 commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,122 @@
+    align-items: var(--toast-align-items, center);
+    margin: var(--toast-margin, 0px 10px 10px 10px);
+    justify-content: space-between;
+    z-index: 1000;
```
have you tested this out with Breeze checkout in Iframe. is it working?

badri-singhal replied:
yes, tested and shared the video with you.

AmreetKumarkhuntia commented on src/lib/Toast/Toast.svelte:
```diff
@@ -0,0 +1,122 @@
+
+  export let properties: ToastProperties = defaultToastProperties;
+  const dispatch = createEventDispatcher();
+  const colorMap: Record<string, string> = {
```
you can add this to constants.ts lets just keep this file for logic

badri-singhal replied:
updated.

AmreetKumarkhuntia commented:
@badri-singhal can you give access for video ?

AmreetKumarkhuntia commented on src/lib/Toast/properties.ts:
```diff
@@ -0,0 +1,21 @@
+  message: 'Toast'
+};
+
+export const colorMap: Record<string, string> = {
```
something to indicate different cases colorMapping

AmreetKumarkhuntia commented on src/lib/Toast/properties.ts:
```diff
@@ -0,0 +1,21 @@
+  duration: number;
+  leftIcon?: string;
+  message: string;
+  showRightIcon?: boolean;
```
Can add option to take in color mapping incase we want to customise it

badri-singhal committed:
BZ-17985: feat: Added Toast Component

- Toast with type success, info, warn and error
- Customizable functionality like message, duration etc.
